### DISCUSSION TITLE
refractory period in a closed loop experiment

### DISCUSSION THREAD

**Discussion by chiyu1203** [bonsai-rx/bonsai#967] [Q&A]
Dear Bonsai team and the lovely community

**the main question still remains but I updated my second questions

It is a little bit embarrassing for me to ask this question as I had similar thing 3 years ago. However, just a slight change in the condition, I am facing another wall. Three years ago, I wanted a refractory period after a stimulus is delivered < https://groups.google.com/g/bonsai-users/c/Za3DqLUkzAQ/m/X6zxEyWZBQAJ> During the refractory period, the condition needs be off first before the next condition is met (this was designed in case the animal had no response to the stimulus and kept the condition on in the entire refractory period in a freely-moving experiment). However, I do not want the second mechanism anymore, meaning that no matter the animal respond or not, as long as its performance met the condition after the refractory period, the stimulus will be delivered. This is basically introducing a minimum length of inter stimulus interval in the paradigm. I tried TimeGate and Gate Interval like this discussion mentioned <https://groups.google.com/g/bonsai-users/c/85sEnWo3f3E/m/VvNT0gcPBgAJ>. 
However it did not work. Could anyone suggest me how to change my original design to fit this?

****
My second question is that the LogEvent made an extra record when I press S to "Subscribe" the experiment (in a sense to start the workflow). This, however, zips with the parameters of the upcoming stimulus. This means the actual timing of the stimulus comes with the parameters of the next event. This sounds like a trivial question, but I still have no clue that where i should start to fix this problem... 

I have attached my workflow below. Feel free to let me know if you have any thought.
Best Regards
Chi-Yu
[refractory_period.zip](https://github.com/bonsai-rx/bonsai/files/9084248/refractory_period.zip)

**Comment by bruno-f-cruz** [comment]
RE: 1. Can you point in your attached workflow where the "During the refractory period, the condition needs be off first before the next condition is met (this was designed in case the animal had no response to the stimulus and kept the condition on in the entire refractory period in a freely-moving experiment)." is implemented? I spent 10min looking around the workflow and I just dont see what you are pointing out. Sorry :/

RE: 2. would simply skip the first element with Skip(1) work?

cheers,
Bruno

**Reply by chiyu1203** [reply]
Hi Bruno,

1. It is in the CreateObservable (from the main Workflow) >>  SelectMany >> "Publish module" The Publish module was connected to a SubscribeWhen, which stop delivering the next stimuli when Publish does not allow. Inside the Publish, I had a Take-DelaySubscription-Repeat module, which set a timeframe to make sure that that the next command can only be relayed outside this timeframe. This is how I set the refractory period.

2. I tried to add Skip (1) behind ReplaySubject (Inside CreateObservable, named Parameters) but it failed. I think the problem with that is because the nature of this workflow. The observable sequence from "Parameters" only output one value each time when a stimulus is delivered. So by adding a Skip(1), I would remove all the output from "Parameters". From my understand about this problem is that the input for the parameters of the visual stimuli is already sent to Parameters-Zip before conditions are met (the stimulus was not delivered due to those "SubscribeWhen" modules), waiting for the next timestamp to arrive (from the TimeStamp > TImeStamp.TimeOfDay). Therefore, to solve this issue, I probably should change all the "SubscribeWhen"-based logic to something else, but I am not sure what options I could explore.

Best Regards

Chi-Yu
